pkg/cache/v2/cache.go

```diff
@@ -145,30 +212,111 @@ func (r *RawResponse) GetDiscoveryResponse() (*discovery.DiscoveryResponse, erro
 	return marshaledResponse.(*discovery.DiscoveryResponse), nil
 }
 
+// GetDeltaDiscoveryResponse performs the marshalling the first time its called and uses the cached response subsequently.
+// We can do this because the marshalled response does not change across the calls.
+// This caching behavior is important in high throughput scenarios because grpc marshalling has a cost and it drives the cpu utilization under load.
+func (r *RawDeltaResponse) GetDeltaDiscoveryResponse() (*discovery.DeltaDiscoveryResponse, error) {
+	marshaledResponse := r.marshaledResponse.Load()
+
+	if marshaledResponse == nil {
+		marshaledResources := make([]*discovery.Resource, len(r.Resources))
+
+		for i, resource := range r.Resources {
+
+			marshaledResource, err := MarshalResource(resource)
+			if err != nil {
+				return nil, err
+			}
+			version, err := HashResource(marshaledResource)
+			if err != nil {
+				return nil, err
+			}
+
+			name := GetResourceName(resource)
+			marshaledResources[i] = &discovery.Resource{
+				Name: name,
+				Resource: &any.Any{
+					TypeUrl: r.DeltaRequest.TypeUrl,
+					Value:   marshaledResource,
+				},
+				Version: version,
+			}
+		}
+
+		marshaledResponse = &discovery.DeltaDiscoveryResponse{
+			Resources:         marshaledResources,
+			RemovedResources:  r.RemovedResources,
+			TypeUrl:           r.DeltaRequest.TypeUrl,
+			SystemVersionInfo: r.SystemVersionInfo,
+		}
+		r.marshaledResponse.Store(marshaledResponse)
+	}
+
+	return marshaledResponse.(*discovery.DeltaDiscoveryResponse), nil
+}
+
 // GetRequest returns the original Discovery Request.
 func (r *RawResponse) GetRequest() *discovery.DiscoveryRequest {
 	return r.Request
 }
 
+// GetDeltaRequest returns the original DeltaRequest
+func (r *RawDeltaResponse) GetDeltaRequest() *discovery.DeltaDiscoveryRequest {
+	return r.DeltaRequest
+}
+
 // GetVersion returns the response version.
 func (r *RawResponse) GetVersion() (string, error) {
 	return r.Version, nil
 }
 
+// GetSystemVersion returns the raw SystemVersion
+func (r *RawDeltaResponse) GetSystemVersion() (string, error) {
+	return r.SystemVersionInfo, nil
+}
+
+// NextVersionMap returns the delta version map built internally by the cache for the state of a snapshot
```

Comment:
update to talk about "next version". This is the downside of all these repeated comments :)

Reply:
Will do, I know sorry about that. I tried to comment as much as possible but the repetitive nature is no good